### PR TITLE
allow mocks to be overwritten

### PR DESCRIPTION
**Pull request by maxgalbu**
Related to https://github.com/padraic/mockery/issues/309

### CODE REVIEW

**Comment by aik099** [comment]
Doesn't sound right because it would leave "allow overwrite" flag and next expectations will replace existing ones all the time if flag isn't removed.

If it's a must have feature (@davedevelopment, what you think of this feature?) then instead I would introduce a method, that would clear previously added expectations when called. Not sure what to do with data recorded (in case of spies), when expectation to be deleted were already matched. We might have some issues with "assertion count" number at the end of the test.

File: library/Mockery/Container.php [hunk 1]
```diff
@@ -67,6 +67,13 @@ class Container
     protected $_loader;
 
     /**
+     * Should we overwrite mocks of the same class?
+     *
+     * @var boolean
```
Comment (GrahamCampbell) [inline library/Mockery/Container.php]: bool not boolean

**Comment by maxgalbu** [comment]
@aik099 yeah sorry should be false by default

**Comment by aik099** [comment]
@maxgalbu , totally missed that `false` thing. But anyway I belive clearing expectations once is better then `enable flag` + `set 1 expectation` + `disable flag`.

**Comment by davedevelopment** [comment]
Could you possibly explain the use case for this?

**Comment by maxgalbu** [comment]
I'm using mockery to mock classes used by other classes, so I don't actually use expectations after mocking. I have a test for a class that checks for user permissions, and I need to change (in the same test) the class that returns those user permission to force a specific permission. Without this change mockery simply adds subsequent mocks in the $this->_mocks array as indexes:

https://github.com/padraic/mockery/blob/3aee65d0a6a9b5e3e8b95a4a4b7830257e7ed608/library/Mockery/Container.php#L419-L423

**Comment by davedevelopment** [comment]
Sorry, I still don't fully understand the use case, could you post some code? Would something like this work?

``` php
$mock = Mockery::mock("overload:SomeClass");

$mock->shouldReceive("getPermissions")->andReturn(644)->byDefault();
var_dump((new SomeClass)->getPermissions()); // 644
$mock->shouldReceive("getPermissions")->andReturn(666)->byDefault();
var_dump((new SomeClass)->getPermissions()); // 666
$mock->shouldReceive("getPermissions")->andReturn(640)->byDefault();
var_dump((new SomeClass)->getPermissions()); // 640
```

See #483

**Comment by maxgalbu** [comment]
It's working for a simple script, but not in the actual test. The only different thing from my test is `byDefault()`

**Comment by davedevelopment** [comment]
I don't understand what you mean, are you saying that is the kind of thing you're looking for?

**Comment by maxgalbu** [comment]
Yes, I probably already tried with your code, but since it wasn't working on my script i looked for another way :D so i made this pull request

**Comment by davedevelopment** [comment]
Cool, so I just merged #483, which should make the `byDefault` work for instance mocks, you'll just need to depend on master until a release gets put together

**Comment by maxgalbu** [comment]
Actually the byDefault was working even without using dev-master from composer, I don't really know what's going on in my tests

**Comment by Squeezya** [comment]
i use codeception for Laravel tests and i use mockery, what i do is 
```
if (self::$myMock) {
    Mockery::close();
}   
self::$myMock = Mockery::mock('alias:myMockClass'); ->with namespace
self::$myMock->shouldReceive('user')->andReturn(new User([]))->byDefault();
```
the close allow to clear the mocks

**Pull request closed without merging**